### PR TITLE
Attempt to fix should be unique bug

### LINKED ISSUE #45: ShouldBeUnique Bug
If a Job implements `ShouldBeUnique` a new `trackedJob` is created even Laravel discarded the "duplicated" Job from the Queue.

Further checks need to be in implemented in the `Trackable` trait constructor before creating the model.

Test Job:
 ```php
class DebugJob implements ShouldQueue, ShouldBeUnique
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;
    use Trackable;

    public function uniqueId(): string
    {
        return $this->trackable->id ?? $this->trackable->uuid;
    }


    public function handle()
    {
        sleep(30);
    }
}
```

Dispatch few times in 30 sec:
```php
    $someModel = SomeModel::first();
    DebugJob::dispatch($someModel);
```

### PR DESCRIPTION
**Pull request by mateusjunges**
This attempts to fix #45

### CODE REVIEW

File: src/Concerns/Trackable.php [hunk 1]
```diff
@@ -27,6 +38,11 @@ public function __construct()
         ]);
     }
 
+    private function isUniqueJobAndCanAcquireLock(): bool
+    {
+        return (new UniqueLock(Container::getInstance()->make(Cache::class)))->acquire($this);
```
Comment (mateusjunges) [inline src/Concerns/Trackable.php:43]: I think this won't work as laravel will acquire this lock before when dispatching this job 🤔

**Comment by sandermanneke** [comment]
@mateusjunges 

Perhaps it is an option to look at the side effects of the job not being created when unique?

I just checked and the second **trackedJob** that is created for the job that has ShouldBeUnique has null for job_id - which makes sense as it does not exist. Why not validate for the presence of a job_id - if null the do not create a tracked job? 

image

**Comment by mateusjunges** [comment]
@sandermanneke that's a good idea. I'll try to implement it this week and see if this fixes the issue. Thanks!

**Comment by DimaVIII** [comment]
@mateusjunges Hey, I have no time currently to work on this but take a look at my "unfinished" approach: https://github.com/DimaVIII/trackable-jobs-for-laravel/tree/features-v2
(It was a while ago and now it's out of sync with your recent work.)

```php
Event::listen(
    \Illuminate\Queue\Events\JobQueued::class,
     \Junges\TrackableJobs\Listeners\QueueTrackedJob::class
);
```

I was going through the Laravel source code of Queue and Horizon and came up with the solution to use event listen, similar to how Horizon handles "should be unique";
It's possible the only way to fix this issue and the best approach to work with the queue system as it opens the possibility to add queue management to this package.

It's a breaking change to the main branch as the database structure was altered to fit the logic, that's why I was originally thinking to release this under v2.

I will add a pull request in case you want implement and extend the logic.

**Comment by mateusjunges** [comment]
Closed this. After giving this some thought, the job being tracked even when not queued should not be considered as a bug, as the intent of this package is to track all your jobs. You can decide how to handle this in your own frontend or even just delete the entry in the database. Thanks for your input on this @DimaVIII!

**Pull request closed without merging**